### PR TITLE
KAFKA-14482: Move LogLoader to storage module

### PR DESCRIPTION
**Pull request by mimaison**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 1]
```diff
@@ -1775,7 +1776,7 @@ class UnifiedLog(@volatile var logStartOffset: Long,
         producerExpireCheck.cancel(true)
         leaderEpochCache.foreach(_.clear())
         val deletedSegments = localLog.deleteAllSegments()
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: maybe `deleteAllSegments` can return java collection directly

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+     * Invokes the provided function and handles any IOException raised by the function by marking the
+     * provided directory offline.
+     *
+     * @param logDirFailureChannel Used to asynchronously handle log directory failure.
+     * @param logDir               The log directory to be marked offline during an IOException.
+     * @param errorMsgSupplier     The supplier for the error message to be used when marking the log directory offline.
+     * @param function             The function to be executed.
+     * @return The value returned by the function after a successful invocation
+     */
+    public static <T> T maybeHandleIOException(LogDirFailureChannel logDirFailureChannel,
```
Comment (chia7712) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java:61]: Could you please replace `LocalLog.scala#maybeHandleIOException`  by java version?

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 3]
```diff
@@ -906,7 +907,7 @@ class UnifiedLog(@volatile var logStartOffset: Long,
               updateHighWatermarkWithLogEndOffset()
 
               // update the producer state
-              updatedProducers.values.foreach(producerAppendInfo => producerStateManager.update(producerAppendInfo))
+              updatedProducers.values.asScala.foreach(producerAppendInfo => producerStateManager.update(producerAppendInfo))
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: how about `updatedProducers.values.forEach`

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 4]
```diff
@@ -1952,19 +1953,17 @@ class UnifiedLog(@volatile var logStartOffset: Long,
...
-      fun
-    }
+    JLocalLog.maybeHandleIOException(logDirFailureChannel, parentDir, () => msg, () => fun)
   }
 
   private[log] def splitOverflowedSegment(segment: LogSegment): List[LogSegment] = lock synchronized {
     val result = UnifiedLog.splitOverflowedSegment(segment, localLog.segments, dir, topicPartition, config, scheduler, logDirFailureChannel, logIdent)
     deleteProducerSnapshots(result.deletedSegments, asyncDelete = true)
-    result.newSegments.toList
+    result.newSegments.asScala.toList
```
Comment (chia7712) [inline core/src/main/scala/kafka/log/UnifiedLog.scala:1962]: it seems the return value is never used, so maybe we can remove it.

**Review by chia7712 — COMMENTED** [review]
@mimaison thanks for nice rewriting.

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 4 above)
Comment (mimaison) [reply core/src/main/scala/kafka/log/UnifiedLog.scala:1962]: The return value is used in a few places in `LogLoaderTest`.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java (hunk 2 above)
Comment (mimaison) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java:61]: This requires a bunch of changes in `LocalLog`. I think it's best if we do those when we move `LocalLog`.

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 3 above)
Comment (mimaison) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: Right we don't need this conversion.

File: core/src/main/scala/kafka/log/UnifiedLog.scala (hunk 1 above)
Comment (mimaison) [reply core/src/main/scala/kafka/log/UnifiedLog.scala]: Good point!

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java (hunk 2 above)
Comment (chia7712) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java:61]: my point was to leverage `LocalLog.java#maybeHandleIOException` to implement `LocalLog.scala#maybeHandleIOException`. That can remove the duplicate code. for example:
```scala
  private def maybeHandleIOException[T](msg: => String)(fun: => T): T = {
    JLocalLog.maybeHandleIOException(logDirFailureChannel, parentDir, () => msg, () => fun)
  }
```

Comment (mimaison) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/LocalLog.java:61]: Ah right, I see what you meant now. Done

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**